### PR TITLE
Basis for Apache module distribution specific abstraction

### PR DESCRIPTION
**Pull request by joohoi**
This PR is mainly restructuring of old methods and Debian/Ubuntu specific methods.

- Started configurator cleanup
- Debian specific code moved to override_debian.py
- Created apache_util.py which hosts generic utility functions that previously resided in configurator
- Added override functionality. The constants.py has map of override classes for each distribution (that requires one), as well as a generic function get_override() which will fetch the correct override class for your distribution and initialize it.

 - Override classes
    -  Classes get reference to the main Configurator element
    -  Methods get caller information {"class": caller_class, "method": caller_method} as first parameter, if they need to call the generic methods as a callback for example.
    - Distribution specific override class will be accessible from configurator

TODO
 - Decide if distribution specific constants should go into override class
 - Decide if override classes should be 1) intercallable 2) dynamically generated, if for example non-debian distribution needs sites-available / sites-enabled functionality

### CODE REVIEW

**Comment by joohoi** [comment]
I considered that approach originally as well, but unfortunately we need overrides for `ApacheParser` too, for example in the cases we recently discussed about ways distributions handle Defines and extra Includes. So deriving only the configurator wouldn't cut it. With this approach it's possible if we simply save `configurator.os_info()` to `parser.os_info()` in parser `__init__()`.

File: certbot-apache/certbot_apache/constants.py [hunk 1]
```diff
@@ -2,6 +2,11 @@
 import pkg_resources
 from certbot import util
 
+from certbot.plugins import common
+
+from certbot_apache import override_debian
+from certbot_apache import override_centos
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/constants.py]: That's completely reasonable, I overlooked them for some reason.

**Comment by bmw** [comment]
Ah good point. When we start overriding parser functionality, were you planning on putting the overridden methods in the same Override class or in a new one for the parser? While it probably doesn't matter too much, I think it's probably best to put it in a separate class as it breaks things up into smaller pieces that are easier to digest and tries to help keep the functionality provided by the configurator and parser distinct.

If we go this route, we could do something like defining a `create_apache_parser` method in the base configurator that returns a new Apache parser. The base configurator could use this method which could be overridden by any distro specific derived classes to use a different parser object.

Thoughts? We definitely don't have to go with this approach and I'm not necessarily opposed to going with what you've done so far. My main concern with the current approach is it is custom code to keep track of the base configurator object and method which gets handled automatically by Python when overriding class methods using inheritance.

**Comment by joohoi** [comment]
The original plan was to use the same override class, as for most of the distros the current need to override functionality is well under five methods total for each. I'll think about that approach a bit.

**Comment by bmw** [comment]
> The original plan was to use the same override class, as for most of the distros the current need to override functionality is well under five methods total for each.

That number might change in the future. (Although then again, it might not.) Even so, my feeling is putting the functionality in separate classes doesn't hurt anything in general, although it does complicate the current approach.

> I'll think about that approach a bit.

Let me know what you think. Again, I'm not necessarily against your approach if there are benefits and we don't have to go with my first proposal. There are other options too. I started looking into class metaprogramming in Python which lead to my comments about the relationship between `object` and `type` in IRC yesterday. Most parts of class metaprogramming are almost certainly overkill, but Python gives us a lot of options here.

**Comment by joohoi** [comment]
> Let me know what you think.
I like it! And with moving parser initialization to a `get_parser()` method, it's easy to override as well.
I created a new PR using this approach, both of these have the same features. The new PR is #5202

**Comment by bmw** [comment]
I'm closing this PR as there is more work on #5202 and I think we both prefer that approach. Please reopen this if there's some reason to take this PR that I overlooked!

**Pull request closed without merging**